src/game_objects.rs

```diff
@@ -44,80 +87,85 @@ pub struct Card {
     pub on_turn_end: Vec<Effect>,
     pub on_strike: Vec<Effect>,
     pub on_defend: Vec<Effect>,
-    pub cost: i16,
-    pub cost_currency: String,
+    pub cost: Cost,
     pub draw_to: DrawTo
 }
 
-pub type CardFactory = HashMap<String, Card>;
-
-#[derive(Serialize, Deserialize, Debug, Clone, Default)]
-pub struct Hand {
-    pub size: usize,
-    pub cards: Vec<Card>
+#[derive(Clone, Debug, Serialize, Deserialize, Default)]
+pub struct CardContainer {
+    pub zone: BoardZone,
+    pub cards: Vec<Card>,
+    pub size: Option<usize>
 }
 
-impl Hand {
+impl CardContainer {
+    pub fn new(zone: BoardZone) -> Self {
+        Self { zone: zone, cards: Vec::<Card>::new(), size: None }
+    }
+
+    pub fn new_sized(zone: BoardZone, size: usize) -> Self {
+        Self { zone: zone, cards: Vec::<Card>::with_capacity(size), size: Some(size) }
+
+    }
+
+    pub fn add(&mut self, card: Card) {
+        self.cards.push(card)
+    }
+
+    pub fn empty(&self) -> bool {
+        self.cards.is_empty()
+    }
+
+    pub fn break_one(&mut self) {
+        self.cards.remove(0);
+    }
+
     pub fn is_full(&self) -> bool {
-        self.cards.len() == self.size
+        match self.size {
+            Some(size) =>  self.cards.len() == size,
+            None => false
+        }
+        
     }
 
     pub fn get(&self, idx: usize) -> Card {
         self.cards[idx].clone()
     }
+
+    fn remove(&mut self, card_idx: usize) -> Card {
+        self.cards.remove(card_idx)
+    }
+
+    pub fn all_effects(&self, event_selector: fn(&Card)-> &Vec<Effect>) -> Vec<(BoardZone, Card, Effect)> {
+
+        self.cards.iter()
+                .flat_map(|c| izip!(iter::repeat(self.zone), 
+                                    iter::repeat(c).cloned(),
+                                    event_selector(c).iter().cloned()) )
+                .collect()
+    }
```

Comment:
This requires a documentation comment (after looking at it, I was quite confused what it actually does).

Suggestion:

```rust
/// Extract a list (iterator) of effects for each card in the container (effects can repeat). 
/// Join these lists together, tagging every effect with the card that causes it and zone the card belongs to.
///
/// Expected use: to evaluate on_turn_start effect etc.
```